### PR TITLE
Add support for math.isclose() and numpy.isclose()

### PR DESCRIPTION
**Pull request by njriasan**
I am implementing support for np.isclose and math.isclose in response to issue #5977. I have currently added support for the numpy implementation but the math implementation has not started yet. My changes add a new function to arraymath.py and four new unit tests. I should have complete support for optional arguments, although there may be issues where I am not doing type conversions that should be allowed.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for submitting this PR to the Numba issue tracker, it's been added to the queue for review. Please be advised however that we are currently in a PR freeze as we are preparing to ship a release candidate (RC). We will begin reviewing PRs again once the RC is out and we thank you in advance for your patience!

**Comment by stuartarchibald** [comment]
Looks like CI is failing on a minor `flake8` violation https://dev.azure.com/numba/numba/_build/results?buildId=6235&view=logs&j=9b1c419d-97bd-5aaf-47ca-b83be1ff9bc1&t=d1ed00e6-cc06-521a-e00c-4f5e4a0f55c5&l=78

**Comment by njriasan** [comment]
Thanks @stuartarchibald. I accidentally made my first commit without flake setup on my machine, which was likely the cause of the issue. I believe both np.isclose() and math.isclose() should be fully working now. The only thing I haven't support is the * keyword to limit the number of positional arguments in math.isclose() (I don't know if numba supports that but its unnecessary and was giving me issues).

No rush on reviewing the PR of course. Good luck with your release and let me know if you have any questions or suggests (for example if I should move my functions somewhere else or change my logic.

**Comment by stuartarchibald** [comment]
@njriasan thanks for fixing the flake8 issue. Don't worry if the implementations have some limitations to begin with, it's entirely normal for Numba's support of a feature to start small and grow. Thanks again for writing the PR :)

File: numba/cpython/mathimpl.py [hunk 1]
```diff
@@ -455,3 +458,26 @@ def gcd(a, b):
...
+        raise TypingError("All arguments must be scalars.")
+    if not isinstance(rel_tol, types.Number) or not isinstance(abs_tol, types.Number):
+        raise TypingError("All arguments must be scalars.")
+
+    def impl(a, b, rel_tol=1e-09, abs_tol=0.0):
+        if a == b:
+            return True
+        abs_diff = np.abs(a - b)
+        min_tolerance = np.maximum(rel_tol * np.maximum(a, b), abs_tol)
+        return abs_diff <= min_tolerance
```
Comment (stuartarchibald) [inline numba/cpython/mathimpl.py:482]: Perhaps take a look at the CPython implementation for how to do this, Numba needs to match this and it'd be a good place to gain insight in to what to additionally consider. https://github.com/python/cpython/blob/8e19c8be87017f6bef8e4c936b1e6ddacb558ad2/Modules/mathmodule.c#L2931-L2971

File: numba/np/arraymath.py [hunk 2]
```diff
@@ -4422,3 +4422,45 @@ def impl(a, b):
         return _cross2d_operation(a_, b_)
 
     return impl
+
+
+#----------------------------------------------------------------------------
+# Equality tests implemented on Floats
+# Implemented based on this feature request
+# - https://github.com/numba/numba/issues/5977
```
Comment (stuartarchibald) [inline numba/np/arraymath.py:4430]: Can remove this.

File: numba/np/arraymath.py [hunk 3]
```diff
@@ -4422,3 +4422,45 @@ def impl(a, b):
...
+    y = np.logical_and(y, np.logical_not(np.isinf(abs_diff)))
+    # Python default behavior is to have NaN comparison be False,
+    # so handle equal_nan separately
+    if equal_nan:
+        # Logically and together NaN between A and B and Or it with y
+        nans = np.logical_and(np.isnan(a), np.isnan(b))
+        y = np.logical_or(y, nans)
+    # Add an extra equality check in case there is an inf
+    equality = a == b
+    return np.logical_or(y, equality)
```
Comment (stuartarchibald) [inline numba/np/arraymath.py:4466]: Be careful to match NumPy's logic, even down to order of operations. e.g. NumPy does `atol + rtol * abs(b)`.
It also shortcuts for the all-finite comparison case. For reference: https://github.com/numpy/numpy/blob/97d2db483fc0ffd46f38d0e1c39d5fc001e33197/numpy/core/numeric.py#L2280-L2316

File: numba/np/arraymath.py [hunk 4]
```diff
@@ -4422,3 +4422,45 @@ def impl(a, b):
...
+# - https://github.com/numba/numba/issues/5977
+@overload(np.isclose)
+def np_isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+    if not type_can_asarray(a) or not type_can_asarray(b):
+        raise TypingError("Inputs a and b must be array-like.")
+    if not isinstance(rtol, types.Float) or not isinstance(atol, types.Float):
+        raise TypingError("Relative and absolute tolerance must " +
+                          "be represented as floats.")
+    if not isinstance(equal_nan, types.Boolean):
+        raise TypingError("Equal_nan must be cast to a bool.")
```
Comment (stuartarchibald) [inline numba/np/arraymath.py:4439]: Can't Numba just call `bool` on `equal_nan`?

File: numba/np/arraymath.py [hunk 5]
```diff
@@ -4422,3 +4422,45 @@ def impl(a, b):
...
+# Equality tests implemented on Floats
+# Implemented based on this feature request
+# - https://github.com/numba/numba/issues/5977
+@overload(np.isclose)
+def np_isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+    if not type_can_asarray(a) or not type_can_asarray(b):
+        raise TypingError("Inputs a and b must be array-like.")
+    if not isinstance(rtol, types.Float) or not isinstance(atol, types.Float):
+        raise TypingError("Relative and absolute tolerance must " +
+                          "be represented as floats.")
```
Comment (stuartarchibald) [inline numba/np/arraymath.py:4437]: I think they just need to be non-complex numbers? e.g. `np.isclose(1, 2, np.uint(3))` works fine. I think something like `if <the type> not in types.real_domain | types.integer_domain:` could work.

**Comment by jpivarski** [comment]
Along with #6857, I might take over this PR if it's dormant. I'll be needing `np.nan_to_num` and `np.isclose`.

Are you planning to finish it, @njriasan?

**Comment by njriasan** [comment]
> Along with #6857, I might take over this PR if it's dormant. I'll be needing `np.nan_to_num` and `np.isclose`.
> 
> Are you planning to finish it, @njriasan?

Feel free to take over this PR. I don't have the time to finish it right now.

**Comment by guilhermeleobas** [comment]
I have an old PR opened which implements all `np.is*` functions, including `np.isclose`:
https://github.com/numba/numba/pull/4610

**Comment by zxweed** [comment]
Hello, is there any chance to merge the `np.isclose()` support?

**Comment by guilhermeleobas** [comment]
Hi @zxweed, I have a PR that is waiting for review
https://github.com/numba/numba/pull/7067

**Comment by guilhermeleobas** [comment]
@njriasan, I am only adding `np.isclose`. Perhaps you can keep this one for `math.isclose`. Also, would you like to review my PR?

**Comment by njriasan** [comment]
> @njriasan, I am only adding `np.isclose`. Perhaps you can keep this one for `math.isclose`. Also, would you like to review my PR?

@guilhermeleobas Sure. I'll take a look at your PR and update this one to correctly implement `math.isclose()`.

**Comment by njriasan** [comment]
To avoid issues with merging this since it is nearly 2 years old, I opened a new PR that actually implements math.isclose properly https://github.com/numba/numba/pull/8199. Closing this PR.

**Pull request closed without merging**